File: Data Structures & Algorithms/permutation-string/submission-0.java

class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if (s1.length() > s2.length()) {
            return false;
        }

        int[] s1Count = new int[26];
        int[] windowCount = new int[26];

        for (char c : s1.toCharArray()) {
            s1Count[c - 'a']++;
        }

        int windowSize = s1.length();

        for (int i = 0; i < windowSize; i++) {
            windowCount[s2.charAt(i) - 'a']++;
        }

        if (matches(s1Count, windowCount)) {
            return true;
        }

        for (int right = windowSize; right < s2.length(); right++) {
            // Add right character
            windowCount[s2.charAt(right) - 'a']++;

            // Remove left character
            int left = right - windowSize;
            windowCount[s2.charAt(left) - 'a']--;

            if (matches(s1Count, windowCount)) {
                return true;
            }
        }

        return false;
    }

    private boolean matches(int[] a, int[] b) {
        for (int i = 0; i < 26; i++) {
            if (a[i] != b[i]) {
                return false;
            }
        }

        return true;
    }
}

File: Data Structures & Algorithms/minimum-window-with-characters/submission-0.py
class Solution(object):
    def minWindow(self, s, t):
        from collections import Counter

        need = Counter(t)
        window = {}

        left = 0
        have = 0
        need_count = len(need)

        result = ""
        result_len = float("inf")

        for right in range(len(s)):
            ch = s[right]

            # Character window mein add karo
            window[ch] = window.get(ch, 0) + 1

            # Required character ki sufficient quantity mil gayi
            if ch in need and window[ch] == need[ch]:
                have += 1

            # Window valid hai
            while have == need_count:

                # Smallest answer save karo
                if right - left + 1 < result_len:
                    result_len = right - left + 1
                    result = s[left:right + 1]

                # Left character remove karo
                left_ch = s[left]
                window[left_ch] -= 1

                if left_ch in need and window[left_ch] < need[left_ch]:
                    have -= 1

                left += 1

        return result
